### PR TITLE
Speeding up decimal (de)serialization.

### PR DESCRIPTION
**Pull request by stebet**
## Proposed Changes

Serializing the `decimal` datatype allocates 40 bytes every time and deserializing it is sub-optimal. This optimization removes that allocation and speeds up (de)serializing decimal values

## Types of Changes

What types of changes does your code introduce to this project?
_Put an `x` in the boxes that apply_

- [ ] Bug fix (non-breaking change which fixes issue #NNNN)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause an observable behavior change in existing systems)
- [ ] Documentation improvements (corrections, new content, etc)
- [ ] Cosmetic change (whitespace, formatting, etc)

## Checklist

- [X] I have read the `CONTRIBUTING.md` document
- [X] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq)
- [X] All tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in related repositories

## Results
``` ini
BenchmarkDotNet=v0.12.1, OS=Windows 10.0.20270
Intel Core i7-10700 CPU 2.90GHz, 1 CPU, 16 logical and 8 physical cores
.NET Core SDK=5.0.101
  [Host]        : .NET Core 3.1.10 (CoreCLR 4.700.20.51601, CoreFX 4.700.20.51901), X64 RyuJIT
  .NET 4.8      : .NET Framework 4.8 (4.8.4261.0), X64 RyuJIT
  .NET Core 3.1 : .NET Core 3.1.10 (CoreCLR 4.700.20.51601, CoreFX 4.700.20.51901), X64 RyuJIT
```
### Before
|         Method |           Job |      Mean |     Error |    StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated | Code Size |
|--------------- |-------------- |----------:|----------:|----------:|-------:|------:|------:|----------:|----------:|
|    DecimalRead |      .NET 4.8 | 21.939 ns | 0.1721 ns | 0.1526 ns |      - |     - |     - |         - |     781 B |
|   DecimalWrite |      .NET 4.8 | 29.016 ns | 0.0880 ns | 0.0735 ns | 0.0063 |     - |     - |      40 B |     686 B |
|    DecimalRead | .NET Core 3.1 |  5.785 ns | 0.0151 ns | 0.0133 ns |      - |     - |     - |         - |     449 B |
|   DecimalWrite | .NET Core 3.1 | 14.726 ns | 0.2515 ns | 0.2353 ns | 0.0048 |     - |     - |      40 B |     300 B |

### After
|       Method |           Job |      Mean |     Error |    StdDev | Gen 0 | Gen 1 | Gen 2 | Allocated | Code Size |
|------------- |-------------- |----------:|----------:|----------:|------:|------:|------:|----------:|----------:|
|  DecimalRead |      .NET 4.8 | 12.192 ns | 0.0406 ns | 0.0360 ns |     - |     - |     - |         - |     592 B |
| DecimalWrite |      .NET 4.8 | 13.177 ns | 0.1229 ns | 0.1089 ns |     - |     - |     - |         - |     659 B |
|  DecimalRead | .NET Core 3.1 |  2.439 ns | 0.0136 ns | 0.0121 ns |     - |     - |     - |         - |     227 B |
| DecimalWrite | .NET Core 3.1 |  3.068 ns | 0.0142 ns | 0.0126 ns |     - |     - |     - |         - |     284 B |

### CODE REVIEW

File: projects/RabbitMQ.Client/client/impl/WireFormatting.cs [hunk 1]
```diff
@@ -34,32 +34,56 @@
...
+    // https://github.com/microsoft/referencesource/blob/master/mscorlib/system/decimal.cs
+    // The lo, mid, hi, and flags fields contain the representation of the
+    // Decimal value. The lo, mid, and hi fields contain the 96-bit integer
+    // part of the Decimal. Bits 0-15 (the lower word) of the flags field are
+    // unused and must be zero; bits 16-23 contain must contain a value between
+    // 0 and 28, indicating the power of 10 to divide the 96-bit integer part
+    // by to produce the Decimal value; bits 24-30 are unused and must be zero;
+    // and finally bit 31 indicates the sign of the Decimal value, 0 meaning
+    // positive and 1 meaning negative.
+    readonly struct DecimalData
```
Comment (bollhals) [inline projects/RabbitMQ.Client/client/impl/WireFormatting.cs:53]: could we move it inside WireFormatting as private class? It's not needed outside.

**Comment by bollhals** [comment]
I assume decimals aren't used as often, but the improvement is substantial nonetheless. 👍

File: projects/RabbitMQ.Client/client/impl/WireFormatting.cs (hunk 1 above)
Comment (stebet) [reply projects/RabbitMQ.Client/client/impl/WireFormatting.cs:53]: Sure.

File: projects/RabbitMQ.Client/client/impl/WireFormatting.cs [hunk 2]
```diff
@@ -34,32 +34,56 @@
...
         public static decimal ReadDecimal(ReadOnlySpan<byte> span)
         {
             byte scale = span[0];
+            ValidateDecimalScale(scale);
+            uint unsignedMantissa = NetworkOrderDeserializer.ReadUInt32(span.Slice(1));
+            var data = new DecimalData(((uint)(scale << 16)) | unsignedMantissa & 0x80000000, 0, unsignedMantissa & 0x7FFFFFFF, 0);
+            return Unsafe.As<DecimalData, decimal>(ref data);
+        }
+
+        private static void ValidateDecimalScale(byte scale)
```
[the remark this answers is not in the document]
Comment (stebet) [reply projects/RabbitMQ.Client/client/impl/WireFormatting.cs:81]: Yeah, that makes sense.

**Comment by stebet** [comment]
> I assume decimals aren't used as often, but the improvement is substantial nonetheless. 👍

I figured since it's a user-provided field, it might be used considerably in some cases (financial apps come to mind), so better to have most primitives as fast as possible.

File: projects/RabbitMQ.Client/client/impl/WireFormatting.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (stebet) [reply projects/RabbitMQ.Client/client/impl/WireFormatting.cs]: Done and results updated.

**Review by michaelklishin — APPROVED** [review]

**Comment by michaelklishin** [comment]
@stebet can you please rebase this on top of the current master?

**Comment by stebet** [comment]
Will do. Might not happen until Monday though.

**Comment by michaelklishin** [comment]
Sure, no rush :)

**Comment by stebet** [comment]
Rebase done.

**Comment by stebet** [comment]
> Sure, no rush :)

Done :)

**Comment by michaelklishin** [comment]
I assume this can be backported all the way to `6.x`?

**Comment by stebet** [comment]
> I assume this can be backported all the way to `6.x`?

Yup :)

**Pull request merged**